src/lib/logical_query_plan/jit_aware_lqp_translator.cpp

```diff
@@ -0,0 +1,294 @@
+
+namespace opossum {
+
+JitAwareLQPTranslator::JitAwareLQPTranslator() : LQPTranslator() {
+#if !HYRISE_JIT_SUPPORT
+  Fail("Query translation with JIT operators requested, but jitting is not available");
+#endif
+}
+
+std::shared_ptr<AbstractOperator> JitAwareLQPTranslator::translate_node(
+    const std::shared_ptr<AbstractLQPNode>& node) const {
+  auto jit_operator = _try_translate_node_to_jit_operators(node);
+  return jit_operator ? jit_operator : LQPTranslator::translate_node(node);
+}
+
+std::shared_ptr<JitOperatorWrapper> JitAwareLQPTranslator::_try_translate_node_to_jit_operators(
+    const std::shared_ptr<AbstractLQPNode>& node) const {
+  uint32_t num_jittable_nodes{0};
+  std::unordered_set<std::shared_ptr<AbstractLQPNode>> input_nodes;
+
+  // Traverse query tree until a non-jittable nodes is found in each branch
+  _visit(node, [&](auto& current_node) {
+    if (_node_is_jittable(current_node)) {
+      ++num_jittable_nodes;
+      return true;
+    } else {
+      input_nodes.insert(current_node);
+      return false;
+    }
+  });
+
+  // It does not make sense to create a JitOperatorWrapper for fewer than 2 LQP nodes,
+  // but we may want a better heuristic here
+  if (num_jittable_nodes < 2 || input_nodes.size() != 1) {
+    return nullptr;
+  }
+
+  // The input_node is not being integrated into the operator chain, but instead serves as the input to the JitOperators
+  const auto input_node = *input_nodes.begin();
+
+  auto jit_operator = std::make_shared<JitOperatorWrapper>(translate_node(input_node));
+  auto read_tuple = std::make_shared<JitReadTuple>();
+  jit_operator->add_jit_operator(read_tuple);
+
+  auto filter_node = node;
+  while (filter_node != input_node && filter_node->type() != LQPNodeType::Predicate &&
+         filter_node->type() != LQPNodeType::Union) {
+    filter_node = filter_node->left_input();
+  }
+
+  // If we can reach the input node without encountering a UnionNode or PredicateNode,
+  // there is no need to filter any tuples
+  if (filter_node != input_node) {
+    // However, if we do need to filter, we first convert the filter node to a JitExpression ...
+    const auto expression = _translate_node_to_jit_expression(filter_node, *read_tuple, input_node);
+    if (!expression) return nullptr;
+    // make sure that the expression gets computed ...
+    jit_operator->add_jit_operator(std::make_shared<JitCompute>(expression));
+    // and then filter on the resulting boolean.
+    jit_operator->add_jit_operator(std::make_shared<JitFilter>(expression->result()));
+  }
+
+  // Identify the top-most projection node to determine the output columns
+  auto top_projection = node;
+  while (top_projection != input_node && top_projection->type() != LQPNodeType::Projection) {
+    top_projection = top_projection->left_input();
+  }
+
+  // Add a compute operator for each computed output column (i.e., a column that is not from a stored table).
+  auto write_table = std::make_shared<JitWriteTuple>();
```

Comment:
reopening: "Naming mismatch: write table vs. write tuple? I think it should be JitWriteTable or JitWriteTuples"